### ISSUE TITLE
🔨 crazy mechanics we don't have yet

### ISSUE THREAD

**Issue by TeoTwawki** [DerpyProjectGroup/topaz#4183]
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my issue will be ignored.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have searched existing [issues](https://github.com/LandSandBoat/server/issues) to see if the issue has already been opened, and I have checked the commit log to see if the issue has been resolved since my server was last updated.

## Describe the feature
As I sat down to write issues per Zach's request on discord, I struggled to come up with actual actionable examples instead of vague memories of wanting xyz in order to do a thing somewhere 3 years (or so) ago. I expect its like that for other people too - they remember something being missing somewhere and stopped trying to implement it, or worked around it in a way that wouldn't find its way up-stream, but any issue for its was several project iterations ago and has gone out of sight out of mind. That unfortunately make sit difficult for anyone to actually implement before running into the instance its needed in.

While not what was wanted, I've still make a short general set here of challenges I've seen halt contributions before. This issue can be closed out and or items crossed off here as better issues are made to replace them. 

Or if such a general list is just to vague to work with, just close it, and we'll try this again another day with better actionable examples to make individual issues with.

- [ ] damage clamping (we don't know how exactly it works, but it doesn't _seem_ to just be -dt/-pdt and that needs confirmed) <br> BCNM 20 Crustacean Conundrum and several other fights use this. Damage gets restricted down to a low number. the scaling changes so that if you are really strong you can get maybe 1 extra damage point, but even ws will be single digit. This needs some retail experimentation. Emperor Arthro makes a good test subject since he isn't level capped and he toggles this state on and off.
- [ ] while we have mods to reduce and increase different types of damage, additional hooks and mixins are needed to make these usable/obvious to the layperson for mobs that swap states during battles on a multitude of conditions: pcs (HTB avatars) during skills/casts, hp percent, during "2 hours",
- [ ] pet damage handling/typing/reduction separately from player damage: <br>one bcnm exists where the mobs resistant to magic are ALSO resistant to bst jug pet melee regardless of slashing/piercing/blunt typing, and the mob resistant to magic in that fight takes full damage from jug pets (Annihilated Anthony) basically the rules for which reductions are in play seem to be different if the attacker is a pet making it problematic to jus use the modifiers we have now
- [ ] similarly to pet damage, need to tell ws and skillchain damage part from other physical/magical damages in a few cases, though I struggle to remember where exactly
- [ ] skill cycling/ordering (funguar "cap" rotation and regeneration) <br>The skills and skill order is dependent on funguar type but even wild normie roaming funguars in la thiene have a set order they will use their skills is, visibly losing a shroom from their cap and preventing re-use of that skill until they reach the final skill in the series that causes their subanimation state to reset and restore the mushrooms. This is actually able to be done in the skill script right now, if someone takes the time to map out which skills for which funguars and what those sub animation values are. **We have some of those done already for some funguars, but not all funguars are identical game wide.** There are some similar styled things on other mobs, with skills that are conditional or ordered but these haven't been documented well. _**steal** can remove a cap, preventing use of a skill, which is a big deal._
- [ ] Methods to adjust the likely hood of a particular skill or spell being selected, based on some conditions such as mob or target HP percentage. <br>Example: bees are less likely use pollen if HP is other very high or very low. they have a greater chance of selecting it around 30% or so. Final sting is more likely below the pollen threshold. Both skills _can_ happen at any percentage, they are just less likely too. This happens with a lot of different skills on many different mobs, and causes some of the difficulty to be different from retail on battlefields and NMs. 
- [ ] smarter mob behavior<br>A much older version of the game had 2 shadow NM in the oubliette beneath San d'Orea with what I refer to as "smart" AI. They didn't random select skills, they used them based on what the player tried to do (you can See ArkEV do similar - she'll shield bash my rdm for trying to cast in range of her). Those 2 NM are now "dumb" and do stuff at random but used to strategically stun healers and sleep the party to interrupt skillchains. Gambits for mobs would make this possible. Or we could hook more player actions for mobs to react to, which might be saner.
- [ ] Mobs that search for other mobs to charm/consume.<br> While we do have [**an** implementation of the latter,](https://github.com/LandSandBoat/server/blob/base/scripts/zones/Abyssea-Konschtat/mobs/Fistule.lua#L10) it is tailored to a very specific case in a way that's not ideal to applying more broadly. And a mob charming another mob is also a case we've not dealt with. We need a reusable thing the layperson can use.
- [ ] knockback "aura" - see [reaving wind](https://ffxiclopedia.fandom.com/wiki/Amphiptere/Video)
- [ ] mobs that cast offensive spells on self, usually to heal (see [Turel](https://ffxiclopedia.fandom.com/wiki/Turul)) we have several ways we could do this, but none readily accessible to the layperson and no clearly established pattern we want people to follow.

**Comment by siknoz** [comment]
If relevant to the list, since you pointing out funguar cap rotation, one thing we're currently missing as well is the ability to steal/despoil multiple items from a mob. These are specifically for the brasscaps/coppercaps/electrumcaps in WOTG areas, though I'm unsure if there's more examples of this behavior. I have two captures showing that you steal items according to their "caps" in order: 

- 3 caps remaining = Steal/despoil Woozyshroom
- 2 caps remaining = Steal  Danceshroom
- 1 cap remaining = Steal Agaricus 
- 0 caps remaining = No item to steal

Also if you successfully steal it also removes the "cap" but despoil does NOT remove the cap. So in theory if you stole 3 times before it used any of the "shroom" TP moves, you would steal a Woozyshroom, a Danceshroom, and an Agaricus from the same mob and it will be unable to use Queasyshroom, Numbshroom, Shakeshroom. 

Captures:
Grauberg - Brasscap - Despoil and Steal same item and Steal 2 Items from same mob
https://youtu.be/7YQmnRheDQY
https://www.dropbox.com/scl/fi/f6dponxx3rkk81a4x64w9/Grauberg-Brasscap-Despoil-and-Steal-same-item-and-Steal-2-Items.zip?rlkey=pwj16cc4ufu1r2mbwnfcn569i&dl=0
Grauberg - Brasscap - Steal 2 items from same mob
https://youtu.be/_d6ZVtWjNqM
https://www.dropbox.com/scl/fi/1nm636qzib8eenqwd5gc6/Grauberg-Brasscap-Steal-2-items.zip?rlkey=90cug78bhpx18v5rbig1p3oc3&dl=0

**Comment by TeoTwawki** [comment]
yes thank you! forgot about that but removing their ability to do a move is a big deal! we can use the hood in the steal script to have the lil buggers set a var on themselves when stolen from and skillcheck can look for that. If you can steal from them more than once they are a rare exception to the normal steal rules and that will also need handled.

**Comment by UmeboshiXI** [comment]
Not sure if this one counts: Mobs that search for other mobs to charm/consume. 

Is there currently a way for a mob to search for other entities in a radius near it without hardcoding a mobid? Maybe having a binding to search for entities near it which can then be narrowed/refined by other existing conditions in lua like mob family, etc.

https://www.bg-wiki.com/ffxi/Category:Rafflesia
Rafflesia can charm bees and then use abilities on them to consume them/hp drain. 

https://www.bg-wiki.com/ffxi/Dextrose
Actively searches for other flans to steal hp from during the battle.

**Comment by TeoTwawki** [comment]
> Not sure if this one counts: Mobs that search for other mobs to charm/consume.

definitely counts, adding to the list

**Comment by UmeboshiXI** [comment]
Thought of some more things as I ran into them. Some mobs have "Enhanced" versions of their normal moves. 

See "Heavy Bellow" for [Nightmare Bugard](https://www.bg-wiki.com/ffxi/Nightmare_Bugard). It is a normally single target turned into an AOE Move. 

We currently have bindings to change spell behavior such as turning normally single target spells into an AOE or changing the AOE distance. Perhaps we could copy these functions to mobskills as well.

Edit: I assume the enhanced skills are usually implemented as different mobskill IDs but it would be convenient to have additional bindings if not for flexibility in handling unique one off skills/support for custom stuff down the line.

**Comment by TeoTwawki** [comment]
> Thought of some more things as I ran into them. Some mobs have "Enhanced" versions of their normal moves.
> 
> See "Heavy Bellow" for [Nightmare Bugard](https://www.bg-wiki.com/ffxi/Nightmare_Bugard). It is a normally single target turned into an AOE Move.
> 
> We currently have bindings to change spell behavior such as turning normally single target spells into an AOE or changing the AOE distance. Perhaps we could copy these functions to mobskills as well.
> 
> Edit: I assume the enhanced skills are usually implemented as different mobskill IDs but it would be convenient to have additional bindings if not for flexibility in handling unique one off skills/support for custom stuff down the line.

with mobskills there is usually a duplicate string with a different ID in the index to use to handle those situations (and actually matches the IDs SE uses, the reason the duplicates exist)
